### ISSUE TITLE
OkHttp 3.3.4-RC1 strange headers behavior with HTTP2 enabled

### ISSUE THREAD

**Issue by DASAR** [square/okhttp#2699]
I have updated from OkHttp 3.3.1 to 3.3.4-RC1 and then have faced strange issue:
When I change request header value became merged with value from previous request (we have saw it on server side). It looks like this SESSION: <oldValue>, <newValue> instead of SESSION: <newValue>

This is exactly what we saw on server side - header has 2 different sessions in it 0ed91** and dc35275*:
`'api-session': '0ed91033efc8950f18c85afa4e6537a3; 7RCbXBqq#zK=Tmfk03hsR7vaRra7jka8S8bMZh0fd1VJkq7?Ydtp=Y&uNNguGvoI; active; 1467802743964; 900000, dc352754a1606f641acac300a9d81adb; 7kch9FjctfxHSQb29NGytrQe7XICtn9WyPtmT2yKJ@s4ou0#Gikg&bvYfY@XxHN6; not-active; 1467802755394; 900000',`

This happens only if Http2 enabled and client use OkHttp3 ver 3.3.4-RC1.

Server: Nginx 1.11.1 (Http2 Enabled)
Client: Nexus 5 (Android 6.0.1)
Connection: HTTPS (TLS1.2)

This is my interceptor used to add session header to each request and get new session header from each response.

```JAVA
Interceptor apiInterceptor = new Interceptor() {
            @Override
            public okhttp3.Response intercept(Chain chain) throws IOException {
                Request original = chain.request();

                Request.Builder builder = original.newBuilder()
                        .method(original.method(), original.body());

                if (model.getSession() != null) {
                    builder.addHeader(SESSION_HEADER, model.getSession().getHeader());
                }

                Request request = builder.build();

                okhttp3.Response response = chain.proceed(request);

                if (response.isSuccessful()) {
                    extractSession(response.headers());
                }

                return response;
            }
        };
```

**Comment by swankjesse** [comment]
Would this be caused my requests executing concurrently?

**Comment by DASAR** [comment]
This SESSION header is set only inside this interceptor per request. I also set some extra headers like Language, Date, API-VERSION, User-Agent.
However response also can contain this header with same name, that's why I extract it from each response to get new value. And I have 2nd Interceptor to do another job related to response.

I think its related to HTTP/2.0, when we test on staging server with HTTP/2 disabled this issue disappears.

**Comment by DASAR** [comment]
I forget to mention app also use okhttp logging-interceptor which is does not show anything wired. Everything seems to work as expected according to logs on client side. Also I have rechecked Model side which is storing and retrieving sessions.

**Comment by swankjesse** [comment]
Could you write a test case to demonstrate the problem? As is it’s hard to understand what we should fix.

**Comment by dave-r12** [comment]
This was likely caused by the bug fixed in #2715. @DASAR can you try version 3.4.1?

This test fails on 3.4.0:
```java
  @Test public void dynamicTableHeaderIndexes() throws IOException {
    hpackWriter.writeHeaders(headerEntries("key", "value", "auth", "secret"));
    bytesIn.writeAll(bytesOut);
    hpackReader.readHeaders();
    assertEquals(headerEntries("key", "value", "auth", "secret"),
        hpackReader.getAndResetHeaderList());

    hpackWriter.writeHeaders(headerEntries("key", "value", "auth", "changedsecret"));
    bytesIn.writeAll(bytesOut);
    hpackReader.readHeaders();
    assertEquals(headerEntries("key", "value", "auth", "changedsecret"),
        hpackReader.getAndResetHeaderList());
  }
```

**Comment by DASAR** [comment]
@dave-r12 @swankjesse with 3.4.1 version this problem disappeared. Everything works as expected.
